constructor/nsis/main.nsi.tmpl

```diff
@@ -456,7 +462,7 @@ Function .onInit
     Push $R2
 
     InitPluginsDir
-
+    @EXTRA_FILES_TO_PLUGINS_DIR@
```

Comment:
Why do we need the files to reside in the plugins dir? Can they be at any other arbitrary location? We could define a "temporary" target directory for `extra_files` so they are not kept in the final installer artifact, but they are indeed located in a known location so the NSI code can locate them. Something like:

```yaml
extra_files:
  my_custom_welcome_image_1.png: "_nsis_temp/"  # or $PluginsDir if that's what we need.
```

Reply:
When the extra files are copied using the `copy_extra_files` in preconda.py, Python has no understanding of NSIS variables like $PluginsDir.  So I think using a temp dir as in your example may work.  The question is what happens in that directory after the install is created.  My hope is that it goes away, if not we need to find a way to designate temporary files.

Reply:
I usually use some variation of `atexit.register(shutil.rmtree, temp_path)` for these cases. I think the only requirement is that the actual temporary path is templated somehow so it's replaced at build time.

Reply:
If this were in Python, I agree with this but the problem we are trying to solve is to not have temporary files just used for installation left around after install. At this point, it is an NSIS problem.  This is why using the NSIS PlugsDir is nice because it is a builtin way to have files that are only used when the installer is run and installed with the files needed to run the application.

Reply:
I feel extra_files should be used for files that will be installed on the target host in the installation process and we want to implement a way label files only meant to be used during the process and not meant to be left on the host.  I am not seeing at clean way to do this with Windows as NSIS is in charge of creating the installer.  I think the cleaner solution would be to make a new keyword that encapsulates the idea of temporary files only meant for installation.

Reply:
Yes, I see your point. The idea was to be "magic" about the target path, by detecting a specific suffix (e.g. `@build-time-only/some_file.txt`). These files wouldn't be copied to the installer, but to a temporary location that would be templated in the NSIS code.

For example, given this `extra_files` config:

```yaml
extra_files:
  - my_image.bmp: @build_time_only/my_image.bmp
```

The template code would have:

```
StrCpy $0 "__BUILD_TIME_ONLY_TEMP_DIR__\my_image.bmp"
System::Call 'user32::LoadImage(i 0, t r0, i ${IMAGE_BITMAP}, i 0, i 0, i ${LR_LOADFROMFILE}|${LR_LOADTRANSPARENT}|${LR_LOADMAP3DCOLORS}) i.s'
```

That, when rendered, would become:

```
StrCpy $0 C:\tmp\constructor_nsfjr7184_tmp\my_image.bmp
System::Call 'user32::LoadImage(i 0, t r0, i ${IMAGE_BITMAP}, i 0, i 0, i ${LR_LOADFROMFILE}|${LR_LOADTRANSPARENT}|${LR_LOADMAP3DCOLORS}) i.s'
```

So the only "contract" is the use of that templated placeholder instead of $PluginsDir.

Is reusing `extra_files` with "magic prefixes" too magic? I am willing to accept a `build_resources` or something like that if the team agrees (but the templated placeholder seems like a good way to go, regardless).

Reply:
@pseudoyim, @marcoesters and I were playing around with this today.  The crux of the problem is that we still need to package the bmp files with the NSIS installer.  The solution above works only locally as it is directly reference the image on the host machine.  We found that temp directory is removed it will not work.

We still need a way to designate to NSIS that the file is temporary.  So I think it is cleaner just to create new keyword.  An alternative approach would be to create code that puts these files in a temporary directory at install and then deletes them but to me this is cleaner:
```
temp_extra_files:
  - my_image.bmp
  ```
  This requires little need for the user to add things to their script as it gets handled.
  
I have a [PR](https://github.com/conda/constructor/pull/594), #594 that has a proposed solution to encapsulate this.

Reply:
Ok, let's go for the new key. About the name, do you think `build_time_files` reflects its purpose better? I am not sure if users would understand that "temporary" here means "temporary at build time" or "temporary at install time".

Reply:
Is 594 supposed to target this PR branch or `main` separately? It looks like it has way too many files in the diff (quick glance only).

Reply:
Yes, @jaimergp , I branched off of this to make and test the changes on #594.  If we went ahead and merged this change the diff would be very small.

As far as the naming goes, I am flexible.  Currently our documentation for `temp_extra_files` is as follows:
```
Temporary files that could be referenced in the installation process (i.e. customized
 `welcome_file` and `conclusion_file` (see above)) . Should be a list of
filepaths, relative to the directory where `construct.yaml` is. In Windows, these
files will be copied into a temporary folder, the NSIS `$PLUGINSDIR`, during
install process (Windows only).
```

I think `build_time_files` may be too specific as there are files that really aren't for building but are used during the installation process.  This is one of our primary use cases at the moment.  I could see this word being expanded to include building resources or files so I think we should let it stand right now as `temp_extra_files`.

Reply:
> I think `build_time_files` may be too specific as there are files that really aren't for building but are used during the installation process. 

Yes, I see what you mean. The intent is that these files will not be part of the final installation, but only required while the installer is being generated. If the installer ends up embedding those resources somehow, that's ok, but the user won't see them in `$INSTDIR` once done, right?

In your current help message "the install process" is not when the user executes the distributed installer, but when the team is building the installer. Is that correct or am I misunderstanding? That's why I was calling them "build-time" files.

My main concern is that `temporary` in here is not a term that really describes how these type of files are being used.

Reply:
Also, can you add the commits of #594 to this PR directly? I'd rather merge this in a single step, and I'll need the whole logic in a single view for the final review :D 

Right now I am a bit confused with `$PluginsDir` and friends 😅

Reply:
Ok, Will do.

Reply:
@jaimergp Should we consider this resolved and close this issue?

Reply:
Sure, I'll add more comments in the current commit if needed. Thanks for the discussion!